jpype/dbapi2.py

```diff
@@ -0,0 +1,1444 @@
+    supported by the database, e.g. requesting a .rollback() on a connection that
+    does not support transaction or has transactions turned off.
+    """
+    pass
+
+
+class _UnsupportedTypeError(InterfaceError, TypeError):
+    pass
+
+
+###############################################################################
+# Connection
+
+_default = object()
+
+
+def connect(dsn, *, driver=None, driver_args=None,
+            adapters=_default, converters=_default,
+            getters=GETTERS_BY_TYPE, setters=SETTERS_BY_TYPE, **kwargs):
+    """ Create a connection to a database.
+
+    Arguments to the driver depend on the database type.
+
+    Args:
+       dsn (str): The database connection string for JDBC.
+       driver (str, optional): A JDBC driver to load.
+       driver_args: Arguments to the driver.  This may either be a dict,
+          java.util.Properties.  If not supplied, kwargs are used as as the
+          parameters for the JDBC connection.
+       *kwargs: Arguments to the driver if not supplied as
+          driver_args.
+
+    Raises:
+       Error if the connection cannot be established.
+
+    Returns:
+       A new connection if successful.
+    """
+    Properties = _jpype.JClass("java.util.Properties")
+    if driver:
+        _jpype.JClass('java.lang.Class').forName(driver).newInstance()
+    DM = _jpype.JClass('java.sql.DriverManager')
+
+    # User is supplying Java properties
+    if isinstance(driver_args, Properties):
+        connection = DM.getConnection(dsn, driver_args)
+
+    # User is supplying a mapping that can be converted Properties
+    elif isinstance(driver_args, typing.Mapping):
+        info = Properties()
+        for k, v in driver_args.items():
+            info.setProperty(k, v)
+        connection = DM.getConnection(dsn, info)
+
+    # User supplied nothing
+    elif driver_args is None:
+        connection = DM.getConnection(dsn)
+
+    # Otherwise use the kwargs
+    else:
+        info = Properties()
+        for k, v in kwargs.items():
+            info.setProperty(k, v)
+        connection = DM.getConnection(url, info)
+
+    return Connection(connection, adapters, converters, setters, getters)
+
+
+class Connection:
+    """ Connection provides access to a JDBC database.
+
+    Connections are managed and can be as part of a Python with statement.
+    Connections close automatically when they are garbage collected, at the
+    end of a with statement scope, or when manually closed.  Once a connection
+    is closed all operations using the database will raise an Error.
+    """
+    Error = Error
+    Warning = Warning
+    InterfaceError = InterfaceError
+    DatabaseError = DatabaseError
+    InternalError = InternalError
+    OperationalError = OperationalError
+    ProgrammingError = ProgrammingError
+    IntegrityError = IntegrityError
+    DataError = DataError
+    NotSupportedError = NotSupportedError
```

Comment:
are these static aliases just for convenience or is there a deeper logic behind this? E.g. downstream you refer to the outer defined exceptions

Reply:
This is a formality of the dbapi2 specification.  I am not sure why it is done but it appears in JayDeBeAPI, MySQL, and Sqlite.   So to conform with other packages I did the same.   My best guess is so that the user can import "Connection" from dbapi2 and then call the errors from the aliases.   But I would have to defer to dbapi2 user.

Reply:
Thanks for the explanation.